### PR TITLE
Added support for incremental image loading

### PR DESCRIPTION
**Pull request by vkvenkat**
Creating incremental containers
Updating data sources 
Getting the container and frame status

### CODE REVIEW

**Comment by jaredhms** [comment]
>         if (index == imageSrc.loadIndex - 1 && index < CGImageSourceGetCount(isrc)) {

comments for some of these checks would be helpful

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:2108 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>     source.loadIndex = index + 1;

comments?

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:1635 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
> size_t CGImageSourceGetCount(CGImageSourceRef isrc) {

can you please add annotations where they're missing?  we should be annotating all APIs that are part of the public surface area.

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:1895 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

File: include/ImageIO/CGImageSourceInternal.h [hunk 1]
```diff
@@ -16,17 +16,29 @@
...
 - (instancetype)initWithDataProvider:(CGDataProviderRef)provider;
+- (instancetype)initIncremental;
 - (CFStringRef)getImageType;
+- (CGImageSourceStatus)getJPEGStatusAtIndex:(size_t)index;
+- (CGImageSourceStatus)getTIFFStatusAtIndex:(size_t)index;
+- (CGImageSourceStatus)getGIFStatusAtIndex:(size_t)index;
+- (CGImageSourceStatus)getBMPStatusAtIndex:(size_t)index;
+- (CGImageSourceStatus)getPNGStatusAtIndex:(size_t)index;
+- (CGImageSourceStatus)getICOStatusAtIndex:(size_t)index;
+- (CGImageSourceStatus)getStatusAtIndex:(size_t)index;
```
Comment (jaredhms) [inline include/ImageIO/CGImageSourceInternal.h]: nit: private/internal methods should have a leading _

**Comment by jaredhms** [comment]
>     static const int c_minDataStreamSize = 96;

nit: already defined at top of file

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:661 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>     static const size_t c_offsetSize = 4;

nit: can you please move these each down to just above where they're first used?  same for the constants in the other helper functions.

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:601 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>         return kCGImageStatusUnknownType;

wouldn't kCGImageStatusUnexpectedEOF be more appropriate for some of these?

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:634 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>         return kCGImageStatusReadingHeader;

can you describe how we'd move past this state?  does the app call again later to see if the image is further along?

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:164 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>         if (offset + 2 > imageLength) {

what is '2'?

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:245 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>         if (offset + 4 > imageLength) {

what is '4'?

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:254 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>         if (offset + 2 > imageLength) {

...in a couple spots

---
In reply to: [217050392](https://github.com/Microsoft/WinObjC/pull/470#issuecomment-217050392) [](ancestors = 217050392)

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:245 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by jaredhms** [comment]
>         if (offset + 4 > imageLength) {

...in a couple spots

---
In reply to: [217050458](https://github.com/Microsoft/WinObjC/pull/470#issuecomment-217050458) [](ancestors = 217050458)

---
Refers to: Frameworks/ImageIO/CGImageSource.mm:254 in 97823c8. [](commit_id = 97823c8402c1fd80959b917153b9bf83c1affd14, deletion_comment = False)

**Comment by vkvenkat** [comment]
CGImageSource.mm:634 - For ICO images, Apple returns kCGImageStatusUnknownType from the minimum stream length of 96 until all of the data is available. The stream length would fall in this category over here.

**Comment by vkvenkat** [comment]
CGImageSource.mm:245, 254 - The succeeding get16BitValue & get32BitValue APIs access the next 2 or 4 bytes beyond the current offset. This is a check to make sure that an out of bounds access doesn't happen here.

**Comment by jaredhms** [comment]
ok, thanks for the clarification.

---
In reply to: [217207673](https://github.com/Microsoft/WinObjC/pull/470#issuecomment-217207673) [](ancestors = 217207673)

**Comment by jaredhms** [comment]
it probably wouldn't hurt - that wasn't immediately clear to me when reading the code

---
In reply to: [217208866](https://github.com/Microsoft/WinObjC/pull/470#issuecomment-217208866) [](ancestors = 217208866)

**Comment by vkvenkat** [comment]
Thanks, added static constants and comments for bound checks

**Pull request merged**